File: interview_question/singleton/singleton.py
from threading import Lock


class Singleton:
    lock = Lock()

    def __new__(cls, *args, **kwargs):
        if not hasattr(cls, '_instance'):
            with cls.lock:
                if not hasattr(cls, '_instance'):
                    cls._instance = object.__new__(cls)
        return cls._instance


if __name__ == '__main__':
    a1 = Singleton()
    a2 = Singleton()
    print(id(a1) == id(a2))

File: interview_question/singleton/singleton.go
package main

import (
	"fmt"
	"sync"
)

type Instance struct {
	name string
}


var once sync.Once
var lock sync.Mutex

// 私有的实例指针
var instance *Instance

// sync.Once版本
func GetInstance() *Instance {
	once.Do(func() {
		instance = &Instance{}	// new(Instance)
	})
	return instance
}

// 带检查锁的单例模式

func GetInstance2() *Instance {

	if instance == nil {
		// 注意加锁的位置，不要一进来就加锁，否则已生成实例后还要加锁，浪费资源
		lock.Lock()
		defer lock.Unlock()
		if instance == nil {
			instance = &Instance{}
		}
	}
	return instance

}

func main() {
	a:=GetInstance2()
	b:=GetInstance2()
	fmt.Printf("%p,%p",a,b)

}